subprojects/core/src/main/java/org/gradle/internal/work/DefaultAsyncWorkTracker.java

```diff
@@ -60,16 +65,23 @@ public void waitForCompletion(Operation operation) {
         }
 
         try {
-            for (AsyncWorkCompletion item : workItems) {
-                try {
-                    item.waitForCompletion();
-                } catch (Throwable t) {
-                    failures.add(t);
-                }
-            }
+            if (workItems.size() > 0) {
+                projectLockService.withoutProjectLock(new Runnable() {
+                    @Override
+                    public void run() {
+                        for (AsyncWorkCompletion item : workItems) {
```

Comment:
It would be nice not to release and reacquire the project lock if everything was already completed.

In a certain sense, these things are just more dependencies for completion of the current operation/task and could potentially be modelled as such.